### ISSUE TITLE
DWA compression randomly ignores big floating-point values.

### ISSUE THREAD

**Issue by cassidyincandela** [AcademySoftwareFoundation/openexr#1319]
Hello there!

This problem involves Blender, so any contributors who also happens to work with Blender that can help, that would be appreciated.

This problem started kind of tame, fixable inside Blender. I rendered a sun disc that emits 100000 max. In this case it's around 70000.
I encode the raw render using DWAA/DWAB.
When I bring those encoded images into Blender's compositor, and did a bunch of operations, I got a region of corrupted pixels.
![Screenshot (2)](https://user-images.githubusercontent.com/118965414/206455108-46b892c5-90d4-4f70-b6d4-50a466bc49f7.png)
Then I find out that the encoded images contain some corrupted pixels, in around the sun.
![Screenshot (1)](https://user-images.githubusercontent.com/118965414/206455444-f0cf18b5-4a7b-4fc5-879c-9e89febb79ed.png)
(the sun is stretched due to motion blur, is expected)

If you see the pixels information below. The corrupted pixels contain the RGB value "inf". I was able to fix this in Blender by mapping the range from infinity to a real number, in this case 65000.

But in the next update of Blender, the encoding is more destructive. High values are ignored instead. When I read the pixels information, it says "-nan(ind)". Not a number. Just like in the first screenshot I provided, except this happens after the OpenEXR export, instead of after compositing. This isn't fixable in Blender because I can't do math operations with no information.

I discussed this problem in more detail in Blender's tracker, because I thought this issue was on Blender's side:
[https://developer.blender.org/T102963](url)

Thanks in advance for any information about this problem.

**Comment by peterhillman** [comment]
The largest value that can be stored in a 16 bit half float is 65504, so the sun cannot be accurately represented. Values above 65504 should be stored as 'infinity', which seems to be happening here. Additionally, DWA lossy compression isn't intended to perform well with very large values, so I'm not surprised that other big values get changed significantly by the compression

Switching to a non-lossy compression for the image may be necessary to handle those bright values better, but will still not represent the sun brightness correctly. You would still need to map infinities to a real number. If blender has an option to map or clamp bright values before writing, you would get better results with DWA compressed 16 bit half float. Perhaps you could also render the entire scene darker so the sun intensity stays below 65504, and brighten it up again afterwards.

If you want to store the sun brightness correctly, you would have to use a 32 bit full float EXR, with a lossless compression scheme such as ZIP, but that will be significantly larger on disk.

**Comment by reunanen** [comment]
I had a similar issue. Tried to clamp values before encoding (e.g., to at most 65504), but they would still come out as infinity after decoding, especially at or near sharp edges where changes in pixel values are steep.

This was happening also when encoding at single precision (and not half).

For me, clamping after decoding (and not before encoding) fixed the problem, but apparently that is not an option for @cassidyincandela.

**Comment by cassidyincandela** [comment]
Actually, that was an option for me haha, I must've forgotten to mention it in the first comment. But I did mention it in [the task I created in Blender's tracker](https://developer.blender.org/T102963), linked at the end of the same comment.

I have abandoned that solution though. The solution that works flawlessly for me now, is to divide the overall values of the image, **before encoding**, until the highest value in the overall image is below `65248` (this is the limit in Blender).

This is the workaround idea from @peterhillman's [comment earlier](#issuecomment-1343407034):
> Perhaps you could also render the entire scene darker so the sun intensity stays below 65504, and brighten it up again afterwards.

Blender's sun disk emits about `3000000` (3 million) in floating-point values so I usually divide my renders by `64`. When opening the image back, I first multiply the overall image by `64`, before doing any other operations. The difference between decoded-clamped and divided-decoded-multiplied images is pretty much nothing, visually, except that dividing-multiplying retains the color of the sun, while clamping would just set the RGB values of the sun to be the same, a.k.a. pure white.

So, I recommend this workaround instead of clamping values.
Here it is for Blender artists who by any chance have stumbled upon this thread and somehow have the exact same circumstances as I do:

![dwa-workaround-in-short](https://user-images.githubusercontent.com/118965414/210070476-5322683b-75c7-46ad-a5b5-8567cde5478f.png)

(Export is encoding, import is decoding.)

I'm going to close this now, since it is considered a normal behavior for DWA codecs, and a satisfying workaround exists.
Thanks for reading and helping folks!
